mcfeely-ttpc.8
''' mcfeely        Asynchronous remote task execution.
''' Copyright (C) 1999 Kiva Networking
'''
''' This program is free software; you can redistribute it and/or
''' modify it under the terms of the GNU General Public License
''' as published by the Free Software Foundation; either version 2
''' of the License, or (at your option) any later version.
'''
''' This program is distributed in the hope that it will be useful,
''' but WITHOUT ANY WARRANTY; without even the implied warranty of
''' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
''' GNU General Public License for more details.
'''
''' You should have received a copy of the GNU General Public License
''' along with this program; if not, write to the Free Software
''' Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
'''
''' You may contact the maintainer at <mcfeely-maintainer@systhug.com>.
.TH mcfeely-ttpc 8
.SH NAME
mcfeely-ttpc \- transer a mcfeely task to a mcfeely client
.SH SYNOPSIS
.B mcfeely-ttpc tasknum
.SH DESCRIPTION
.B mcfeely-ttpc
reads a McFeely task from stdin and makes a TTP2 connection to a
.B mcfeely-ttpd
process on the destination client. The first thing
it reads is the hostname. It uses the mcfeely hosts file to find
the "real host" and port number for connecting to the remote
host. The TTP2 shared secret is read and passed to the
.B mcfeely-ttpd.
Without waiting to see if the secret is accepted
(the remote end will just disregard us if it's not)
.B mcfeely-ttpc
sends down the task number, UID and GID (not currently used) and
the task comm and arguments.

.B mcfeely-ttpc
then waits for results from 
.B mcfeely-ttpd,
gathering the exit code and any output execution of the comm may have
produced. The output is sent to stdout and
.B mcfeely-ttpc
exits indicating success, failure or deferral.
.B mcfeely-spawn
picks up the output and the exit code in it's reaper routine.
.SH "SEE ALSO"
mcfeely-ttpd(8), mcfeely-spawn
.SH "COPYRIGHT"
This program is copyright 2000 Kiva Networking. It is free software;
you can redistribute it and/or modify it under the terms of the GNU
General Public License as published by the Free Software Foundation;
either version 2 of the License, or (at your option) any later version.
Please see the file COPYING in the distribution.
.SH "AUTHORS"
Matt Liggett <mml@pobox.com>, Chris Dent <cdent@systhug.com>,
Adrian Hosey <ahosey@systhug.com>


HISTORY
A Brief History of McFeely and how it got its name
By Adriant J.C. Liggett-McFeely

This software is named in honor of my uncle's brother's 
great-great-grandfather Otis J. McFeely, a recluse millionaire that funded 
my software schooling and adult book store while you learned programming 
techniques. I felt so indebted to him that my first big software undertaking 
was named after him.

Sadly, after a breakfast of Chardonnay and donuts with sprinkles, Otis
stood up, gathered a packet of Burpee apple seeds, a can of Sterno and
his trusty walking stick. Mumbling something about the luck of the
Irish, he wandered off into the bush of Indianapolis to never be heard
from again.

To this day, we still suspect that he roams the Earth, at least in
spirit.



mcfeely-ttpc.c
/* mcfeely        Asynchronous remote task execution.
** Copyright (C) 1999 Kiva Networking
**
** This program is free software; you can redistribute it and/or
** modify it under the terms of the GNU General Public License
** as published by the Free Software Foundation; either version 2
** of the License, or (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
**
** You may contact the maintainer at <mcfeely-maintainer@systhug.com>.
*/

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <netdb.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <unistd.h>
#include <knetstring.h>

#include "hostport.h"

#define TASKID (argv[1])

void   ok(char *m, int l) { write(1, m, l); exit(  0); }
void soft(char *m, int l) { write(1, m, l); exit( 99); }
void hard(char *m, int l) { write(1, m, l); exit(100); }

void soft_write(void) { soft("write failure", 13); }
void soft_read(void)  { soft("read failure",  12); }

/* given a taskid as a string, open the file on fd */
int task_open(taskid)
char *taskid;
{
    char *it;
    int len;
    int fd;

    /* "queue/task/TASKID" -- TASKID+12 */
    len = strlen(taskid) + 12;
    it = (char *)malloc(len);
    if (! it) return 0;

    snprintf(it, len, "queue/task/%s", taskid);
    fd = open(it, O_RDONLY);
    free(it);
    return fd;
}

int
secret_open(fnam)
char *fnam;
{
    char *fn;

    fn = malloc(strlen(fnam)+17); /* "control/secrets/" == 16 */
    strcpy(fn, "control/secrets/");
    strcat(fn, fnam);

    return open(fn, O_RDONLY);
}


void
main(argc, argv)
int argc;
char *argv[];
{
    int pfd;
    knsbuf_t buf = {0, 0, 0};
    char host[1024];
    char *i;
    char *realhost;
    char *secretfile;
    int realport;
    int s;
    struct hostent *ent;
    struct protoent *tcp;
    struct sockaddr_in sin;
    unsigned int tasknum;
    char code;

    bzero(&sin, sizeof(sin));

    /* tcp socket */
    tcp = getprotobyname("tcp");
    s = socket(AF_INET, SOCK_STREAM, tcp->p_proto);

    i = host;
    do {
        if (read(0, i, 1) != 1) soft("unexpected eof", 14);
    } while (*i++ != '\0');

    /* get real hostname and port */
    if (! hostport(&realhost, &realport, &secretfile, host))
        soft("cannot find host in control/hosts", 33);

    /* lookup in DNS */
    ent = gethostbyname(realhost);
    if (!ent) soft("cannot find host", 16);
    if (!ent->h_addr) soft("host has no address", 19);
    
    /* open secret file */
    pfd = secret_open(secretfile);
    if (pfd == -1) soft("cannot open secret", 18);

    /* prepare sockaddr_in */
    bcopy(ent->h_addr, &sin.sin_addr, ent->h_length);
    sin.sin_family = AF_INET;
    sin.sin_port = htons(realport);

    /* connect */
    if (connect(s, (struct sockaddr *)&sin, sizeof(sin)) == -1)
        soft("cannot connect", 14);

    /* send: auth, taskid, UID, GID, the task */
    if (knsfwrite(s, pfd) == -1)      soft_write();
    close(pfd);
    tasknum = strtoul(TASKID, 0, 10);
    if (write(s, &tasknum, 4) != 4)   soft_write();
    if (write(s, "\0\0\0\0", 4) != 4) soft_write();
    if (write(s, "\0\0\0\0", 4) != 4) soft_write();
    if (knsfwrite(s, 0) == 1)       soft_write();

    /* recv: response */
    if (read(s, &code, 1) != 1)  soft_read();
    if (knsbread(s, &buf) == -1) soft_read();

    switch (code) {
        case 'K': ok(buf.start, buf.len);   break;
        case 'Z': soft(buf.start, buf.len); break;
        case 'F': hard(buf.start, buf.len); break;
        default:  hard("garbled report", 14);
    }
}
